### PR TITLE
Add Command Parser feature

### LINKED ISSUE #15: [User Story Feature] Add shell-like Command Parser
## User Story
As a Computing Student, I want to  type in commands that feel familiar to what my local shell (e.g. `bash`, `cmd`, `zsh`) offers, so that I can intuitively pick up WellNus++ and be more productive with it out of the box

## Describe the solution you'd like
Commands typed into the terminal should follow how most shells process commands to make the CLI interface as intuitive and 'close to home' as possible.  

It should use `--` as delimiters to separate parameters from payloads.  

This models how most shells (and CLI programs) process commands. For example,

```bash
git tag --annotate <annotation_message> 
```

Separates the input into the following Argument-Payload pairs
`git, tag`
`annotate, <annotation_message>`

Formally, in WellNus this is the expected behaviour:
```bash
command payload1 --argument2 <payload2: can be empty> --argument3 <payload3: can be empty> ...
```
Should split to
- `command, payload`
- `argument2, payload2`
- `argument3, payload3`

### Design Justification
This architecture provides a few benefits design-wise.
 
1. The common `--` delimiter for ALL arguments makes the design [PREDICTABLE](https://uxdworld.com/2021/07/12/why-predictability-is-important-for-ux-design/) (good design practice) 
2. The argument name that comes right after `--` is descriptive and serves as a natural [SIGNIFIER](https://www.interaction-design.org/literature/topics/signifiers) for the user, so that they will be less confused and the experience will feel more natural
3. Better typing experience. It is far more natural for the user to type in non-alphabetical letters '-' all at one shot before typing the alphabetical body argument

## Describe alternatives you've considered
Command structures such as  [AB3](https://se-education.org/addressbook-level3/UserGuide.html) where input types are specified , e.g. `n/John Doe` are more 'secure' from the get go. However, due to the following issues, AB3 was not chosen as the alternative solution compared to the shell-like structure.

### Steep learning curve
For experienced and inexperienced users, it is a hassle to remember what letter corresponds to what argument.

For AB3, the user needs to remember all the different character 'verbs' such as `e/` for email, `n/` for name

### Does not scale well
AB3 structure runs the high risk of **argument-space collision** as well. 

For example, consider a command that needs an `email` and `entry`. What does `e/<payload>` correspond to? 
We could simply just put entry as some other character -- but that defeats the purpose of having the structure in the first place as the character is the argument's first character.

This makes behaviour unpredictable and a confusing user experience.

### Bad expert user experience
For expert users and CLI-masters, pedantic argument input like AB3 makes the typing experience MUCH slower due to the need to type <character><slash><word> which is relatively clunky as the user will need to type far off to the '/' key on the keyboard.

### PR DESCRIPTION
**Pull request by nichyjt**
This closes #15 .

### Changes
1. Added a `command` package to WellNus
2. Added CommandParser logic which processes string representing user input
- `parseuserInput` logically divides it into argument-payload pairs via a `HashMap` 
- `getMainArgument` provides a useful utility for any caller to get the main argument (action) from a command input string
3. Added in unit tests for public CommandParser methods, accounting for many adversarial test cases

### CODE REVIEW

File: src/test/java/seedu/duke/command/CommandParserTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
+package seedu.duke.command;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class CommandParserTest {
+
```
Comment (haoyangw) [inline src/test/java/seedu/duke/command/CommandParserTest.java:10]: Extra line?

File: src/main/java/seedu/duke/command/BadCommandException.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package seedu.duke.command;
+
+public class BadCommandException extends Exception{
```
Comment (haoyangw) [inline src/main/java/seedu/duke/command/BadCommandException.java:3]: Missing space between 'Exception' and '{'?

**Review by haoyangw — COMMENTED** [review]
Just some thoughts based on reading the code, will test the logic later.

**Review by haoyangw — COMMENTED** [review]
Other than minor style issues, no bugs found in code(mental tracing) and added test class runs without errors. Resolve the minor style issues, and it'll LGTM!

File: src/main/java/seedu/duke/command/CommandParser.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            }
+            currentCommand = currentCommand.strip();
+            // Case 2 check
+            if (currentCommand.startsWith(UNPADDED_DELIMITER)) {
+                throw new BadCommandException(ERROR_EMPTY_COMMAND);
+            }
+            cleanCommands[i] = currentCommand;
+        }
+        return cleanCommands;
+    }
```
Comment (wenxin-c) [inline src/main/java/seedu/duke/command/CommandParser.java:76]: For case 1, what if the argument is totally missing? It fulfills none of the cases and will be mis-treated as a command instead in the subsequent processes?

**Review by wenxin-c — COMMENTED** [review]
Overall good work. The major part need to be extra careful of is the splitIntoCommand method as there may be possible ways to bypass the checking. The rest is just some minor issues which do not affect functionality.

File: src/test/java/seedu/duke/command/CommandParserTest.java (hunk 1 above)
Comment (nichyjt) [reply src/test/java/seedu/duke/command/CommandParserTest.java:10]: You mean line 10?

File: src/main/java/seedu/duke/command/CommandParser.java (hunk 3 above)
Comment (nichyjt) [reply src/main/java/seedu/duke/command/CommandParser.java:76]: Can you give an example?   

Your phrasing "missing argument" is vague as it could mean that the delimiter is present but the name of the argument is not.
Or are you asking if there is no argument at all?  e.g. `mainCommand payload`

If it helps, I  accounted for various edge cases in the [unit test](src/test/java/seedu/duke/command/CommandParserTest.java) that encompass "missing argument"  with inputs such as: 
- `mainCommand payload --`
- `mainCommand payload -- `
- `mainCommand payload -- payload1`
^ All of the above will throw an error as the argument is 'missing'.

File: src/main/java/seedu/duke/command/CommandParser.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+ * A CommandParser processes user input from a defined format <p>
+ * <p><br>
+ * Each user input via console consists of: <br>
+ * 1. COMMANDS - A Argument and Payload pairs <br>
+ * 2. ARGUMENTS - String representing the action/parameters of the command <br>
+ * 3. PAYLOADS - value of the action/parameters <br>
+ * In short, user input is a list of commands, each command containing arguments and payloads. <br>
+ * <br>
+ * For example, a given user input: <code>"deadline work on CS2113 --by Sunday"</code>
+ * <li>Has commands ["deadline work on CS2113", "by Sunday"]</li>
```
[the remark this answers is not in the document]
Comment (nichyjt) [reply src/main/java/seedu/duke/command/CommandParser.java:15]: Ok, will be rewriting the documentation with a quick explanation that `" --"` is used as a delimiter to separate argument-payload pairs

File: src/main/java/seedu/duke/command/CommandParser.java (hunk 3 above)
Comment (wenxin-c) [reply src/main/java/seedu/duke/command/CommandParser.java:76]: let say:
```
mainCommand payload --payload1
```
there is no space between '--' and 'payload1'
so what might have in this case?

Comment (nichyjt) [reply src/main/java/seedu/duke/command/CommandParser.java:76]: Oh! This is actually defined behaviour.  

In this case, payload1 is accepted as an argument (key value).  

CommandParser's job is to screen for syntactically wrong inputs only. Setting payload as argument is a logical error on the users' part.   

As we discussed, the responsibility of logic-based checks like type-checking and length-checking is up to each feature using CommandParser to avoid tightly coupling the Parser to each feature.

This also follows the [Single Responsibility Principle](https://nus-cs2113-ay2223s2.github.io/website/se-book-adapted/common/definitions.html#s).   

A benefit of the hashmap architecture is that we can effectively ignore bad inputs like these.  
We can just `.get` the keys we expect to have.

If we want to be super strict, on the feature level, once an unknown key is encountered, the feature can just throw a custom `InvalidArgument` exception or handle the error along those lines. 

I don't see a way to differentiate arguments from payloads without each feature's parsing logic being tightly coupled with each feature's unique requirements. What are your thoughts?

Comment (wenxin-c) [reply src/main/java/seedu/duke/command/CommandParser.java:76]: okay it makes sense. Maybe can just leave it to each feature to check for validity of arguments.

File: src/test/java/seedu/duke/command/CommandParserTest.java (hunk 1 above)
Comment (haoyangw) [reply src/test/java/seedu/duke/command/CommandParserTest.java:10]: yeah

Comment (nichyjt) [reply src/test/java/seedu/duke/command/CommandParserTest.java:10]: ok! will fix

File: src/main/java/seedu/duke/command/CommandParser.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ * as the main command
+ */
+public class CommandParser {
+
+    private static final String ARGUMENT_DELIMITER = " --";
+    private static final String UNPADDED_DELIMITER = "--";
+    private static final String PAYLOAD_DELIMITER = " ";
+
+    // Message string constants for errors and ui
+    private static final String ERROR_EMPTY_COMMAND = "Command length is empty.";
```
[the remark this answers is not in the document]
Comment (nichyjt) [reply src/main/java/seedu/duke/command/CommandParser.java:30]: Will be changing this to `"Command is empty!"`

**Review by wenxin-c — APPROVED** [review]
LGTM!!
